storage: add closed timestamp subsystem

tbg opened the pull request:
NB: This is a reopen of #27595 which needs a fix to the CI stress tester
to merge before it can pass CI (and I don't want to wait for that). No
need to review.

----

This commit introduces the basic closed timestamp subsystem. It does
not hook it up into the CockroachDB node yet.

Great care has been taken to make each individual component self-
contained and cleanly interfaced to the outside world. The package
layout relative to `closedts` is the following:

.: slim root package containing only the most basic CT interfaces
./ctpb: protobuf definitions and service
./config: fat package pulling in everything needed to actually run
./storage: storage for per-node closed timestamp state
./minprop: minimum proposal timestamp tracker
./provider: the Provider, centralizing all closed timestamp information.

Most of these packages depend on `closedts` (more precisely, on the
interfaces they need to use), but in turn don't depend on each other.

The following diagram together with the `ctconfig` and `ct` packages
should be instructive in getting a bird's eye view of how the subsystem
fits into CockroachDB: All of the components are grouped in a Container
(intended as a pass-around per-instance Singleton) and is passed down to
the Stores via the StoreConfig. Replicas proposing commands talk to the
Tracker; replicas trying to serve follower reads talk to the Provider,
which receives closed timestamp updates for the local node and its
peers.

```
                            Node 1 | Node 2
                                   |
+---------+  Close  +-----------+  |  +-----------+
| Tracker |<--------|           |  |  |           |
+-----+---+         | +-------+ |  |  | +-------+ |  CanServe
      ^             | |Storage| |  |  | |Storage| |<---------+
      |             | --------+ |  |  | +-------+ |          |
      |Track        |           |  |  |           |     +----+----+
      |             | Provider  |  |  | Provider  |     | Follower|
      |             +-----------+  |  +-----------+     | Replica |
      |                 ^                  ^            +----+----+
      |                 |Subscribe         |Notify           |
      |                 |                  |                 |
+---------+             |      Request     |                 |
|Proposing| Refresh +---+----+ <----- +----+-----+  Request  |
| Replica |<--------| Server |        |  Clients |<----------+
+---------+         +--------+ -----> +----------+  EnsureClient
                                 CT
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27922)
<!-- Reviewable:end -->

tbg committed:
storage: add closed timestamp subsystem

This commit introduces the basic closed timestamp subsystem. It does
not hook it up into the CockroachDB node yet.

Great care has been taken to make each individual component self-
contained and cleanly interfaced to the outside world. The package
layout relative to `closedts` is the following:

.: slim root package containing only the most basic CT interfaces
./ctpb: protobuf definitions and service
./config: fat package pulling in everything needed to actually run
./storage: storage for per-node closed timestamp state
./minprop: minimum proposal timestamp tracker
./provider: the Provider, centralizing all closed timestamp information.

Most of these packages depend on `closedts` (more precisely, on the
interfaces they need to use), but in turn don't depend on each other.

The following diagram together with the `ctconfig` and `ct` packages
should be instructive in getting a bird's eye view of how the subsystem
fits into CockroachDB: All of the components are grouped in a Container
(intended as a pass-around per-instance Singleton) and is passed down to
the Stores via the StoreConfig. Replicas proposing commands talk to the
Tracker; replicas trying to serve follower reads talk to the Provider,
which receives closed timestamp updates for the local node and its
peers.

```
                            Node 1 | Node 2
                                   |
+---------+  Close  +-----------+  |  +-----------+
| Tracker |<--------|           |  |  |           |
+-----+---+         | +-------+ |  |  | +-------+ |  CanServe
      ^             | |Storage| |  |  | |Storage| |<---------+
      |             | --------+ |  |  | +-------+ |          |
      |Track        |           |  |  |           |     +----+----+
      |             | Provider  |  |  | Provider  |     | Follower|
      |             +-----------+  |  +-----------+     | Replica |
      |                 ^                  ^            +----+----+
      |                 |Subscribe         |Notify           |
      |                 |                  |                 |
+---------+             |      Request     |                 |
|Proposing| Refresh +---+----+ <----- +----+-----+  Request  |
| Replica |<--------| Server |        |  Clients |<----------+
+---------+         +--------+ -----> +----------+  EnsureClient
                                 CT
```

Release note: None

tschottdorf commented:
bors r=nvanbenschoten

nvanbenschoten commented:
Awesome!
